File: videoplatform/src/main/java/com/videoplatform/Moderator.java
package com.videoplatform;

import com.google.inject.Inject; // Додаємо імпорт для @Inject

/**
 * Клас Moderator реалізує роль модератора у системі.
 * Демонструє наслідування від User та реалізацію Moderatable інтерфейсу.
 */
public class Moderator extends User implements Moderatable {
    // Приватні атрибути - інкапсуляція
    private int blockedCount;
    private String department;

    /**
     * Конструктор для ініціалізації модератора
     */
    @Inject // Додаємо анотацію @Inject для конструктора
    public Moderator(String name, String email, String department) {
        super(name, email);
        this.department = department;
        this.blockedCount = 0;
    }

    /**
     * Реалізація методу для блокування спаму
     */
    @Override
    public void blockSpam(String content) {
        this.blockedCount++;
        notifySystem("Заблокував спам: \"" + content + "\"");
    }

    /**
     * Реалізація методу для модерації чату
     */
    @Override
    public void moderateChat(String message) {
        System.out.println("[МОДЕРАЦІЯ ЧАТУ]: " + name + " перевіряє повідомлення: \"" + message + "\"");
    }

    /**
     * Реалізація методу для видалення неприпустимих коментарів
     */
    @Override
    public void deleteInappropriateComment(String comment) {
        notifySystem("Видалив неприпустимий коментар: \"" + comment + "\"");
    }

    /**
     * Реалізація абстрактного методу performAction
     */
    @Override
    public void performAction() {
        System.out.println("[ДІЯ МОДЕРАТОРА]: " + name + " модерує контент та чат");
    }

    /**
     * Геттер для отримання назви відділу модератора
     */
    public String getDepartment() {
        return department;
    }

    /**
     * Геттер для отримання кількості заблокованого контенту
     */
    public int getBlockedCount() {
        return blockedCount;
    }
}

File: src/Premiere/Moderator.cs
using System;

namespace Premiere
{
    /// <summary>
    /// Клас Moderator реалізує роль модератора у системі.
    /// Демонструє наслідування від User та реалізацію IModeratable інтерфейсу.
    /// </summary>
    public class Moderator : User, IModeratable
    {
        // Приватні атрибути - інкапсуляція
        private int BlockedCount { get; set; }
        private string Department { get; set; }

        /// <summary>
        /// Конструктор для ініціалізації модератора
        /// </summary>
        public Moderator(string name, string email, string department) : base(name, email)
        {
            Department = department;
            BlockedCount = 0;
        }

        /// <summary>
        /// Реалізація методу для блокування спаму
        /// </summary>
        public void BlockSpam(string content)
        {
            BlockedCount++;
            NotifySystem($"Заблокував спам: \"{content}\"");
        }

        /// <summary>
        /// Реалізація методу для модерації чату
        /// </summary>
        public void ModerateChat(string message)
        {
            Console.WriteLine($"[МОДЕРАЦІЯ ЧАТУ]: {Name} перевіряє повідомлення: \"{message}\"");
        }

        /// <summary>
        /// Реалізація методу для видалення неприпустимих коментарів
        /// </summary>
        public void DeleteInappropriateComment(string comment)
        {
            NotifySystem($"Видалив неприпустимий коментар: \"{comment}\"");
        }

        /// <summary>
        /// Реалізація абстрактного методу PerformAction
        /// </summary>
        public override void PerformAction()
        {
            Console.WriteLine($"[ДІЯ МОДЕРАТОРА]: {Name} модерує контент та чат");
        }

        /// <summary>
        /// Метод для отримання назви відділу модератора
        /// </summary>
        public string GetDepartment()
        {
            return Department;
        }

        /// <summary>
        /// Метод для отримання кількості заблокованого контенту
        /// </summary>
        public int GetBlockedCount()
        {
            return BlockedCount;
        }
    }
}
